lib/resty/rediscluster.lua

```diff
@@ -347,35 +376,66 @@
 end
 
 
-local function pick_node(self, serv_list, slot, magic_radom_seed)
+local function pick_node(self, serv_list, magic_radom_seed, select_ip)
     local host
     local port
     local slave
     local index
+
     if #serv_list < 1 then
-        return nil, nil, nil, "serv_list for slot " .. slot .. " is empty"
+        return nil, nil, nil, "serv_list is empty"
     end
+
     if self.config.enable_slave_read then
         if magic_radom_seed then
             index = magic_radom_seed % #serv_list + 1
         else
             index = math.random(#serv_list)
         end
+
+        -- compatible check if Redis 7 cluster created with IP directly
+        if not serv_list[index].ip then
+            index = index - 1
+        end
```

Comment:
Can you elaborate in the comment why with Redis 7, the node index needs to be decremented if it was created with "IP directly"?

Reply:
I have reformatted the code. It is now more clear and much more simpler than before.